dataland-user-service/src/main/kotlin/org/dataland/datalanduserservice/service/PortfolioBulkDataRequestService.kt

```diff
@@ -0,0 +1,135 @@
+                .filter { filterFunction(it.sector) }
+                .map { it.companyId }
+                .toSet()
+
+        private fun getFinancialsCompanyIds(portfolio: EnrichedPortfolio) =
+            getCompanyIdsBySector(portfolio) {
+                it != null && it.lowercase() == "financials"
+            }
+
+        private fun getUndefinedCompanyIds(portfolio: EnrichedPortfolio) = getCompanyIdsBySector(portfolio) { it == null }
+
+        private fun getNonFinancialsCompanyIds(portfolio: EnrichedPortfolio) =
+            getCompanyIdsBySector(portfolio) { it != null && it.lowercase() != "financials" }
+
+        private fun getAllCompanyIds(portfolio: EnrichedPortfolio): Set<String> = portfolio.entries.map { it.companyId }.toSet()
+
+        private fun getMonitoringPeriods(startingMonitoringPeriod: String?): Set<String> {
+            val startingMonitoringPeriodInt =
+                startingMonitoringPeriod?.toIntOrNull()
+                    ?: throw IllegalArgumentException("Invalid start year: '$startingMonitoringPeriod' is not a valid number")
+            return (startingMonitoringPeriodInt until 2025)
+                .map { it.toString() }
+                .toSet()
+        }
+
+        private fun sendBulkDataRequest(
+            enrichedPortfolio: EnrichedPortfolio,
+            monitoringPeriods: Set<String>,
+            selector: (EnrichedPortfolio) -> Set<String>,
+            types: Set<BulkDataRequest.DataTypes>,
+        ) {
+            val companyIds = selector(enrichedPortfolio)
+            if (companyIds.isEmpty()) {
+                return
+            }
+
+            requestControllerApi.postBulkDataRequest(
+                BulkDataRequest(
+                    userId = enrichedPortfolio.userId,
+                    companyIdentifiers = selector(enrichedPortfolio),
+                    dataTypes = types,
+                    reportingPeriods = monitoringPeriods,
+                    notifyMeImmediately = false,
+                ),
+            )
+        }
+
+        private fun sendFinancialBulkDataRequest(
+            enrichedPortfolio: EnrichedPortfolio,
+            monitoringPeriods: Set<String>,
+        ) = sendBulkDataRequest(
+            enrichedPortfolio, monitoringPeriods, ::getFinancialsCompanyIds,
+            setOf(
+                BulkDataRequest.DataTypes.eutaxonomyMinusFinancials,
+                BulkDataRequest.DataTypes.nuclearMinusAndMinusGas,
+            ),
+        )
+
+        private fun sendNonFinancialBulkDataRequest(
+            enrichedPortfolio: EnrichedPortfolio,
+            monitoringPeriods: Set<String>,
+        ) = sendBulkDataRequest(
+            enrichedPortfolio, monitoringPeriods, ::getNonFinancialsCompanyIds,
+            setOf(
+                BulkDataRequest.DataTypes.eutaxonomyMinusNonMinusFinancials,
+                BulkDataRequest.DataTypes.nuclearMinusAndMinusGas,
+            ),
+        )
+
+        private fun sendUndefinedBulkDataRequest(
```

Comment:
```suggestion
        private fun sendUndefinedBulkDataRequest(
```
undefined = without sector? that might be misguiding